### ISSUE TITLE
Support for Resize shape to fit text

### ISSUE THREAD

**Issue by Wenfengcheng** [ShapeCrawler/ShapeCrawler#360]
_SCAutoFitType.Resize_ exists in autofitType of **TextFrame**, but it is not implemented.
Add the implementation of _ResizeShapeSize_ like the _ResizeFontSize_ method.

**Comment by Wenfengcheng** [comment]
@adamshakhabov The solution for _measureText_ is probably: **SkiaSharp**, **Graphics**, **TextRenderer**.Which one is expected to be used, or other solutions.
[Spire.Presentation](https://www.e-iceblue.com/Tutorials/Spire.Presentation/Spire.Presentation-Program-Guide/Paragraph-and-Text/Auto-Fit-Text-or-Shape-in-PowerPoint-in-C-VB.NET.html) and [Aspose.Slides](https://docs.aspose.com/slides/net/manage-autofit-settings/) also implement the _ResizeShapeSize_ functions, I don't know which one they use.

**Comment by adamshakhabov** [comment]
@Wenfengcheng, thank you for the information.

Some time ago, I researched these Spire.Presentation and Aspose.Slides libraries to find out what they use to convert pptx to image, measureText, etc, but unfortunately, they have good obfuscation🙂 However, yes, I suspect that Aspose.Slide uses SkiaSharp.

**Comment by Wenfengcheng** [comment]
@adamshakhabov I have tried all three methods: **SkiaSharp**, **Graphics**, **TextRenderer**, but the results are quite different from those calculated automatically by the **PowerPoint** client.
Maybe there is something wrong with my calculation logic.

**Comment by adamshakhabov** [comment]
> the results are quite different from those calculated automatically by the PowerPoint client.

Do you mean measureText result?

It's expected. For example, I made a lot of experiments with converting slide into image via Spire, Aspose and other libraries. I got never result like PowerPoint's output. For example, color contrast or font can be different. In other words, PowerPoint's output was always better.

**Comment by Wenfengcheng** [comment]
> Do you mean measureText result?

Yes, have you tried measureText? I try to calculate size using **Graphics** like below, but the results are quite different.

```csharp
public static SizeF GetTextSize(string text)
{
    SizeF sf;

    using (System.Drawing.Font f = new System.Drawing.Font("Arial Black", 14))
    {
        using (var image = new Bitmap(1, 1))
        using (var g = Graphics.FromImage(image))
        {
            StringFormat format = StringFormat.GenericTypographic;
            format.FormatFlags |= StringFormatFlags.MeasureTrailingSpaces;
            sf = g.MeasureString(text, f, new PointF(0, 0), format);
        }
    }
    
    return sf;
}
```

**Comment by adamshakhabov** [comment]
I did some text measuring for `SCAutoFitType.Shrink`: https://github.com/ShapeCrawler/ShapeCrawler/blob/e126aab69983cd2f1259e1ff42d713df7425490f/ShapeCrawler/Services/FontService.cs#L11

but I'm not expert in 2D Graphic and it is my first experience😶

**Comment by adamshakhabov** [comment]
@Wenfengcheng, thank you for the code snippet. It helped me!

I made a base implementation of processing _Resize shape when paragraph text is changed_ flag. I tested this on simple cases. 

You can use this build [**ShapeCrawler.0.37.0-beta.2**](https://github.com/ShapeCrawler/ShapeCrawler/files/9961544/ShapeCrawler.0.37.0-beta.2.zip). If it failed on your use cases feel free to re-open actual issue or open a new one. We will fix it🙂

**Comment by Wenfengcheng** [comment]
@adamshakhabov Thx, I will try this way.

**Comment by Wenfengcheng** [comment]
@adamshakhabov The **AutofitType** for **TextFrame** in **IAutoShape** is **ReadOnly**, maybe it's a read-write property.

```csharp
public interface ITextFrame
{
    IParagraphCollection Paragraphs { get; }

    string Text { get; set; }

    SCAutoFitType AutofitType { get; }

    double LeftMargin { get; }

    double RightMargin { get; }

    double TopMargin { get; }

    double BottomMargin { get; }

    bool CanChangeText();
}
```

**Comment by Wenfengcheng** [comment]
Your current **Resize** logic is to only adjust the height when the width is not enough, but the implemention of the PowerPoint Client is to adjust the width and height according to the text and text property at the same time, maybe we should follow this way.

```csharp
SKRect val2 = default(SKRect);
string text = TextFrame!.Text;
val.MeasureText(text, ref val2);
float width = ((SKRect)(ref val2)).get_Width();
float textSize = val.get_TextSize();
int num5 = base.Width - num - num2;
int num6 = base.Height - num3 - num4;
if (!((float)num5 > width))
{
  float num7 = width / (float)num5;
  int num8 = (int)num7;
  if (num7 - (float)num8 > 0f)
  {
    num8++;
  }
  float num9 = (float)num8 * textSize + (float)num3 + (float)num4;
  if ((float)num6 < num9)
  {
    base.Height = (int)num9 + num3 + num4 + num3 + num4;
    float num10 = (num9 - (float)num6) / 2f;
    base.Y -= (int)num10;
  }
}
```

**Comment by adamshakhabov** [comment]
> @adamshakhabov The **AutofitType** for **TextFrame** in **IAutoShape** is **ReadOnly**, maybe it's a read-write property. For an existing shape, I cannot modify any of its properties except **Text**.

The following issues have been created to achieve this:

- https://github.com/ShapeCrawler/ShapeCrawler/issues/385
- https://github.com/ShapeCrawler/ShapeCrawler/issues/384

**Comment by Wenfengcheng** [comment]
@adamshakhabov For a new **ResizeToFit Shape**, maybe the implementation is no problem. But for a existing **NoFit Shape**, we want to change _AutoFitType_ to **ResizeToFit**, the PowerPoint client app will both update the width and height according to the text inside shape.

https://user-images.githubusercontent.com/30426178/201245116-138d3720-c3b5-44ca-bce4-fc91a07f8a86.mp4

**Comment by adamshakhabov** [comment]
Got it! Will check this case.

**Comment by adamshakhabov** [comment]
> Your current Resize logic is to only adjust the height when the width is not enough, but the implemention of the PowerPoint Client is to adjust the width and height according to the text and text property at the same time, maybe we should follow this way.

@Wenfengcheng, but your code also changes only Height and Y, no **Width**. Or I miss something?

**Comment by Wenfengcheng** [comment]
> > Your current Resize logic is to only adjust the height when the width is not enough, but the implemention of the PowerPoint Client is to adjust the width and height according to the text and text property at the same time, maybe we should follow this way.
> 
> @Wenfengcheng, but your code also changes only Height and Y, no **Width**. Or I miss something?

The code above is decompiled based on **ShapeCrawler.0.37.0-beta.2**, which is your current implementation.

**Comment by adamshakhabov** [comment]
@Wenfengcheng, the setter for `ITextFrame.AutofitType` has been added in [**v0.38.0-beta.1**](https://github.com/ShapeCrawler/ShapeCrawler/files/10096147/ShapeCrawler.0.38.0-beta.1.zip), which also updates height.

**Comment by Wenfengcheng** [comment]
@adamshakhabov it worked perfectly!!! But the default _AutofitType_ value got from **TextFrame** not correctly.
<img width="212" alt="image" src="https://user-images.githubusercontent.com/30426178/205830937-2c411cde-9545-47f4-b734-0989b1b5f210.png">
<img width="522" alt="image" src="https://user-images.githubusercontent.com/30426178/205831105-9a5e57ab-1754-4d3a-9809-f23fc6095dec.png">

**Comment by BryanCWijaya** [comment]
Hi @ashahabov, I think the AutofitType.Resize still not support TextBox with multiple paragraphs (or text with newline), since SkiaSharp doesn't measure the newline. Do you mind if I create a PR to fix this?

**Comment by ashahabov** [comment]
Hello @BryanCWijaya. It would be awesome! Feel to create a PR.

**Comment by BryanCWijaya** [comment]
Great! But could you please tell me how to get this value? @ashahabov 
![image](https://github.com/user-attachments/assets/0d5adf59-5017-442d-8dba-724ffbb559ef)

**Comment by ashahabov** [comment]
> I see LineSpacingLines and LineSpacingPoints but cannot understand what that variable is.

(1) `LineSpacingLines`
(2) `LineSpacingPoints`

![image](https://github.com/user-attachments/assets/e823d8ac-c9da-40b7-ab13-bd91acffbb46)
